File: Models/Connection.cs
using System;
using Avalonia;
using Avalonia.Media;
using Avalonia.Controls;

namespace TeachingAidMac.Models
{
    public class Connection
    {
        public enum ConnectionVisualState
        {
            Normal,
            Highlighted,      // Yellow - final path
            RedHighlighted,   // Red - being explored
            GreenHighlighted, // Green - best path
            OrangeHighlighted // Orange - alternative path
        }

        public Node Source { get; private set; }
        public Node Target { get; private set; }
        public int Weight { get; set; }
        public int PreviousWeight { get; set; }
        public bool PixelMode { get; set; }
        public ConnectionVisualState VisualState { get; set; } = ConnectionVisualState.Normal;

        private Point _startPoint;
        private Point _endPoint;
        private Canvas? _canvas;

        public Node StartNode => Source;
        public Node EndNode => Target;
        public int EdgeNumber => Weight;

        public Connection(Node source, Node target, int weight)
        {
            Source = source;
            Target = target;
            Weight = weight;
            PreviousWeight = 0;
            PixelMode = false;
        }

        public void SetPixelMode(bool pixelMode)
        {
            PixelMode = pixelMode;
            if (pixelMode)
            {
                PreviousWeight = Weight;
            }
            else
            {
                if (PreviousWeight != 0)
                {
                    Weight = PreviousWeight;
                }
            }
            UpdateEdgeNumber();
        }

        public void UpdateEdgeNumber()
        {
            if (PixelMode)
            {
                // Calculate distance between nodes in pixels
                var dx = Source.Bounds.Center.X - Target.Bounds.Center.X;
                var dy = Source.Bounds.Center.Y - Target.Bounds.Center.Y;
                var distance = Math.Sqrt(dx * dx + dy * dy);
                Weight = (int)Math.Round(distance / 20.0);
            }
        }

        public void DrawLine(Canvas? canvas = null)
        {
            _canvas = canvas;
            UpdatePoints();
            // Drawing will be handled by the GraphCanvas control
        }

        public void RemoveLine()
        {
            // Line removal will be handled by the GraphCanvas control
        }

        private void UpdatePoints()
        {
            var sourcePos = Source.GetPosition();
            var targetPos = Target.GetPosition();

            _startPoint = new Point(
                sourcePos.X + Source.Width / 2,
                sourcePos.Y + Source.Height / 2
            );
            _endPoint = new Point(
                targetPos.X + Target.Width / 2,
                targetPos.Y + Target.Height / 2
            );
        }

        public Point GetMidpoint()
        {
            UpdatePoints();
            return new Point(
                (_startPoint.X + _endPoint.X) / 2,
                (_startPoint.Y + _endPoint.Y) / 2
            );
        }

        public Point GetStartPoint()
        {
            UpdatePoints();
            return _startPoint;
        }

        public Point GetEndPoint()
        {
            UpdatePoints();
            return _endPoint;
        }

        public Node GetEndNode()
        {
            return Target;
        }

        public Node GetStartNode()
        {
            return Source;
        }

        public override string ToString()
        {
            return $"{Source.NodeName} -> {Target.NodeName} (Weight: {Weight})";
        }

        public override bool Equals(object? obj)
        {
            if (obj is Connection other)
            {
                return (Source == other.Source && Target == other.Target) ||
                       (Source == other.Target && Target == other.Source);
            }
            return false;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Source, Target);
        }

        // Additional methods for VB compatibility
        public void DrawHighlightedLine()
        {
            Console.WriteLine($"DEBUG: DrawHighlightedLine called for {Source.NodeName}->{Target.NodeName}, Object ID: {this.GetHashCode()}");
            VisualState = ConnectionVisualState.Highlighted;
            Console.WriteLine($"DEBUG: VisualState set to {VisualState}");

            // For bidirectional graphs, also highlight the reverse connection
            var reverseConnection = Target.GetConnectionTo(Source);
            if (reverseConnection != null && reverseConnection != this)
            {
                Console.WriteLine($"DEBUG: Also highlighting reverse connection {Target.NodeName}->{Source.NodeName}, Object ID: {reverseConnection.GetHashCode()}");
                reverseConnection.VisualState = ConnectionVisualState.Highlighted;
            }

            // Trigger graph redraw if available
            if (Source.CurrentGraph != null)
            {
                Console.WriteLine($"DEBUG: Calling TriggerRedraw for {Source.NodeName}->{Target.NodeName}");
                Source.CurrentGraph.TriggerRedraw();
            }
            else
            {
                Console.WriteLine($"DEBUG: Source.CurrentGraph is null for {Source.NodeName}->{Target.NodeName}");
            }
        }

        public void DrawRedHighlightedLine()
        {
            VisualState = ConnectionVisualState.RedHighlighted;

            // For bidirectional graphs, also highlight the reverse connection
            var reverseConnection = Target.GetConnectionTo(Source);
            if (reverseConnection != null && reverseConnection != this)
            {
                reverseConnection.VisualState = ConnectionVisualState.RedHighlighted;
            }

            // Don't auto-trigger redraw to avoid performance issues during exploration
        }

        public void DrawGreenHighlightedLine()
        {
            VisualState = ConnectionVisualState.GreenHighlighted;

            // For bidirectional graphs, also highlight the reverse connection
            var reverseConnection = Target.GetConnectionTo(Source);
            if (reverseConnection != null && reverseConnection != this)
            {
                reverseConnection.VisualState = ConnectionVisualState.GreenHighlighted;
            }

            // Don't auto-trigger redraw to avoid performance issues
        }

        public void DrawOrangeHighlightedLine()
        {
            VisualState = ConnectionVisualState.OrangeHighlighted;

            // For bidirectional graphs, also highlight the reverse connection
            var reverseConnection = Target.GetConnectionTo(Source);
            if (reverseConnection != null && reverseConnection != this)
            {
                reverseConnection.VisualState = ConnectionVisualState.OrangeHighlighted;
            }

            // Don't auto-trigger redraw to avoid performance issues
        }

        public void ResetHighlighting()
        {
            VisualState = ConnectionVisualState.Normal;

            // For bidirectional graphs, also reset the reverse connection
            var reverseConnection = Target.GetConnectionTo(Source);
            if (reverseConnection != null && reverseConnection != this)
            {
                reverseConnection.VisualState = ConnectionVisualState.Normal;
            }

            // Don't trigger redraw here to avoid circular dependency
        }
    }
}

File: Connection.vb
Public Class Connection 

    Private StartNode As Node 

    Private EndNode As Node 

    Private EdgeNumber As Integer 

    Private PreviousEdgeNumber As Integer 

    Private StartPoint As Point 

    Private EndPoint As Point 

 

    Private PixelMode As Boolean 

 

    Public Sub New() 

 

    End Sub 

    Public Sub SetPixelMode(PixelMode As Boolean) 

        Me.PixelMode = PixelMode 

        ' sets the edge number based on the pixel mode 

        If PixelMode = True Then 

            Me.PreviousEdgeNumber = Me.EdgeNumber 

        Else 

            If Me.PreviousEdgeNumber = 0 Then 

 

            Else 

                Me.EdgeNumber = Me.PreviousEdgeNumber 

            End If 

 

        End If 

        UpdateEdgeNumber() 

    End Sub 

    Public Sub UpdateEdgeNumber() 

        If PixelMode = True Then 

            ' sets the edge number to the distance between the nodes in pixels 

            Me.EdgeNumber = 1 * Math.Round(Math.Sqrt(Math.Pow((Me.StartNode.Location.X - Me.EndNode.Location.X), 2) + Math.Pow((Me.StartNode.Location.Y - Me.EndNode.Location.Y), 2)) / 20, 1) 

        End If 

    End Sub 

 

    Public Sub DrawLine() 

 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.Black, 4) 

 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

        'draws line between the midpoint of the startnode and the midpoint of the endnode 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

        g.DrawString(Me.EdgeNumber, SystemFonts.DefaultFont, Brushes.Black, mp.X - 5, mp.Y - 8) 

        ' draws a rectangle at the midpoint of the connection line and draws the edge number in it 

        ' this rectangle expans depending on the edgenumber 

 

    End Sub 

    Public Sub DrawHighlightedLine(StartNode As Node, EndNode As Node) 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.Yellow, 4) 

 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

        'draws line between the midpoint of the startnode and the midpoint of the endnode 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

        g.DrawString(Me.EdgeNumber, SystemFonts.DefaultFont, Brushes.DarkBlue, mp.X - 5, mp.Y - 8) 

        ' draws a rectangle at the midpoint of the connection line and draws the edge number in it 

        ' this rectangle expans depending on the edgenumber 

    End Sub 

    Public Sub DrawRedHighlightedLine() 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.PaleVioletRed, 4) 

 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

        'draws line between the midpoint of the startnode and the midpoint of the endnode 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

        g.DrawString(Me.EdgeNumber, SystemFonts.DefaultFont, Brushes.Black, mp.X - 5, mp.Y - 8) 

        ' draws a rectangle at the midpoint of the connection line and draws the edge number in it 

        ' this rectangle expans depending on the edgenumber 

    End Sub 

    Public Sub DrawGreenHighlightedLine() 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.LightGreen, 4) 

 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

        'draws line between the midpoint of the startnode and the midpoint of the endnode 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

        g.DrawString(Me.EdgeNumber, SystemFonts.DefaultFont, Brushes.Black, mp.X - 5, mp.Y - 8) 

        ' draws a rectangle at the midpoint of the connection line and draws the edge number in it 

        ' this rectangle expans depending on the edgenumber 

    End Sub 

    Public Sub DrawOrangeHighlightedLine() 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.OrangeRed, 4) 

 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

        'draws line between the midpoint of the startnode and the midpoint of the endnode 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

        g.DrawString(Me.EdgeNumber, SystemFonts.DefaultFont, Brushes.Black, mp.X - 5, mp.Y - 8) 

        ' draws a rectangle at the midpoint of the connection line and draws the edge number in it 

        ' this rectangle expans depending on the edgenumber 

    End Sub 

    Public Sub RemoveLine() 

 

        StartPoint = StartNode.Location 

        StartPoint.X += Me.StartNode.Width / 2 

        StartPoint.Y += Me.StartNode.Height / 2 

        EndPoint = EndNode.Location 

        EndPoint.X += Me.StartNode.Width / 2 

        EndPoint.Y += Me.StartNode.Height / 2 

        Dim g As Graphics = Form1.CreateGraphics() 

        Dim pen As New Pen(Color.White, 4) 

 

        g.DrawLine(pen, StartPoint, EndPoint) 

 

        Dim mp As Point = GetMidpoint(StartPoint, EndPoint) 

 

        g.FillRectangle(Brushes.White, mp.X - 5, mp.Y - 8, 10 * Me.EdgeNumber.ToString().Length, 16) 

 

        ' draws over the line and the rectangle with the color of the form 

 

 

    End Sub 

 

 

    Private Function GetMidpoint(p1 As Point, p2 As Point) As Point 

        Dim x As Integer = (p1.X + p2.X) / 2 

        Dim y As Integer = (p1.Y + p2.Y) / 2 

        Return New Point(x, y) 

    End Function 

 

 

    Public Sub SetStartNode(StartNode As Node) 

        Me.StartNode = StartNode 

    End Sub 

    Public Sub SetEndNode(EndNode As Node) 

        Me.EndNode = EndNode 

    End Sub 

    Public Sub SetEdgeNumber(EdgeNumber As Integer) 

        Me.EdgeNumber = EdgeNumber 

    End Sub 

    Public Sub SetSpecificPixelMode(PixelMode As Boolean) 

        Me.PixelMode = PixelMode 

    End Sub 

    Public Sub SetPreviousEdgeNumber(PreviousEdgeNumber As Integer) 

        Me.PreviousEdgeNumber = PreviousEdgeNumber 

    End Sub 

 

    Public Function GetStartNode() As Node 

        Return Me.StartNode 

    End Function 

    Public Function GetEndNode() As Node 

        Return Me.EndNode 

    End Function 

    Public Function GetEdgeNumber() As Integer 

        Return Me.EdgeNumber 

    End Function 

    Public Function GetPixelMode() As Boolean 

        Return Me.PixelMode 

    End Function 

    Public Function GetPreviousEdgeNumber() As Integer 

        Return Me.PreviousEdgeNumber 

    End Function 

End Class 